Example:

```cpp
  TF_ASSIGN_OR_RETURN(a, foo(10));
```

In GCC, it will be expanded into:

```cpp
auto _status_or_value0 = (foo(10)); if (!_status_or_value0.ok()) { return _status_or_value0.status(); } a = std::move(_status_or_value0.ValueOrDie());
```

But in MSVC, it will be expanded into:

```cpp
(a, foo(10));
```

With warning C4003: not enough actual parameters for macro 'TF_STATUS_MACRO_GET_VARIADIC_IMPL'

You can see in https://stackoverflow.com/questions/47752606/msvc-preprocessor-bug , StackOverflower said the original implementation is not standard-compliant. It works in GCC because GCC allows `__VA_ARGS__` to be empty as a compiler extension. MSVC does not allow that.

The new implementation works in GCC as well. Can I just remove the old implementation and use the new one in all compilers?

--- a/tensorflow/compiler/xla/status_macros.h
+++ b/tensorflow/compiler/xla/status_macros.h
@@ -196,34 +196,8 @@ class StatusAdaptorForMacros {
 #define TF_STATUS_MACROS_CONCAT_NAME(x, y) TF_STATUS_MACROS_CONCAT_IMPL(x, y)
 #define TF_STATUS_MACROS_CONCAT_IMPL(x, y) x##y
 
-#ifdef _MSC_VER
-
-#define TF_ASSIGN_OR_RETURN(lhs, rexpr)                                   \
-  TF_ASSIGN_OR_RETURN_IMPL(TF_STATUS_MACROS_CONCAT_NAME(_status_or_value, \
-                                                        __COUNTER__),     \
-                           lhs, rexpr)
-
-#define TF_ASSIGN_OR_RETURN_IMPL(statusor, lhs, rexpr) \
-  auto statusor = (rexpr);                             \
-  if (TF_PREDICT_FALSE(!statusor.ok())) {              \
-    return statusor.status();                          \
-  }                                                    \
-  lhs = std::move(statusor.ValueOrDie())
-
-#else
-
-#define TF_ASSIGN_OR_RETURN(...)                                             \
-  TF_STATUS_MACRO_GET_VARIADIC_IMPL(__VA_ARGS__, TF_ASSIGN_OR_RETURN_IMPL_3, \
-                                    TF_ASSIGN_OR_RETURN_IMPL_2)              \
-  (__VA_ARGS__)
-
-#define TF_STATUS_MACRO_GET_VARIADIC_IMPL(_1, _2, _3, NAME, ...) NAME
-
-#define TF_ASSIGN_OR_RETURN_IMPL_2(lhs, rexpr) \
-  TF_ASSIGN_OR_RETURN_IMPL_3(lhs, rexpr)
-
-#define TF_ASSIGN_OR_RETURN_IMPL_3(lhs, rexpr) \
-  TF_ASSIGN_OR_RETURN_IMPL(                    \
+#define TF_ASSIGN_OR_RETURN(lhs, rexpr) \
+  TF_ASSIGN_OR_RETURN_IMPL(             \
       TF_STATUS_MACROS_CONCAT_NAME(_status_or_value, __COUNTER__), lhs, rexpr)
 
 #define TF_ASSIGN_OR_RETURN_IMPL(statusor, lhs, rexpr) \
@@ -232,7 +206,5 @@ class StatusAdaptorForMacros {
     return statusor.status();                          \
   }                                                    \
   lhs = std::move(statusor.ValueOrDie())
-
-#endif
 
 #endif  // TENSORFLOW_COMPILER_XLA_STATUS_MACROS_H_

